app/Singletons/ChatFilters.php

```diff
@@ -5,24 +5,68 @@
 
 namespace App\Singletons;
 
+use App\Exceptions\ContentModerationException;
 use App\Models\ChatFilter;
 use App\Traits\Memoizes;
 
 class ChatFilters
 {
     use Memoizes;
 
+    /**
+     * Applies all active chat filters to the provided text.
+     * @param string $text The text to filter.
+     * @return string The text after filtering.
+     * @throws ContentModerationException If the text matches one of the filters which indicate that the input
+     * should be discarded entirely.
+     */
     public function filter(string $text): string
     {
-        $replacements = $this->memoize(__FUNCTION__, function () {
+        $filters = $this->memoize(__FUNCTION__, function () {
             $ret = [];
             foreach (ChatFilter::all() as $entry) {
-                $ret[$entry->match] = $entry->replacement;
+                $ret[$entry->match] = $entry;
             }
 
             return $ret;
         });
 
-        return strtr($text, $replacements);
+
+        // blocking filters (finding any of these phrases throws moderation exceptions)
+        $blockingFilters = array_where($filters, fn ($filter) => $filter->block);
+        if (!empty($blockingFilters) && preg_match(self::combinedFilterRegex($blockingFilters), $text)) {
+            throw new ContentModerationException();
+        }
+
+        // non-blocking filters (phrases are allowed to be replaced)
+        $replaceFilters = array_where($filters, fn ($filter) => !$filter->block);
+
+        $whitespaceFilters = array_where($replaceFilters, fn ($filter) => $filter->whitespace_delimited);
+        $nonWhitespaceFilters = array_where($replaceFilters, fn ($filter) => !$filter->whitespace_delimited);
+
+        $text = strtr($text, array_reduce($nonWhitespaceFilters, function ($map, $filter) {
+            $map[$filter->match] = $filter->replacement;
+            return $map;
+        }, []));
+        return preg_replace(
+            array_map(fn ($filter) => self::singleFilterRegex($filter), $whitespaceFilters),
+            array_map(fn ($filter) => $filter->replacement, $whitespaceFilters),
+            $text
+        );
+    }
+
+    public static function singleFilterRegex(ChatFilter $filter): string
+    {
+        $term = preg_quote($filter->match);
+        if ($filter->whitespace_delimited) {
+            $term = '\b'.$term.'\b';
+        }
+        return '('.$term.')';
```

Comment:
Yeah so that was a "I was moving the parens from one function to the other but then apparently blacked out midway through and forgot to actually do the thing to the end" type deal. I think https://github.com/ppy/osu-web/pull/11181/commits/db8dc7c0a3ba840eab11d4be0f9bee883baa05f3 should make this better?